modules/network-load-balancer/main.tf

```diff
@@ -0,0 +1,87 @@
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+# DEPLOY A NETWORK LOAD BALANCER
+# This module deploys a GCP Regional Network Load Balancer
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+terraform {
+  # This module has been updated with 0.12 syntax, which means it is no longer compatible with any versions below 0.12.
+  required_version = ">= 0.12"
+}
+
+# ------------------------------------------------------------------------------
+# CREATE FORWARDING RULE
+# ------------------------------------------------------------------------------
+
+resource "google_compute_forwarding_rule" "default" {
+  provider              = google-beta
+  project               = var.project
+  name                  = var.name
+  target                = google_compute_target_pool.default.self_link
+  load_balancing_scheme = "EXTERNAL"
+  port_range            = var.port_range
+  ip_address            = var.ip_address
+  ip_protocol           = var.protocol
+
+  labels = var.custom_labels
+}
+
+# ------------------------------------------------------------------------------
+# CREATE TARGET POOL
+# ------------------------------------------------------------------------------
+
+variable "instances" {
+  description = "List of self links to instances in the pool. Note that the instances need not exist at the time of target pool creation."
+  type        = list(string)
+  default     = []
+}
+
+resource "google_compute_target_pool" "default" {
+  project          = var.project
+  name             = "${var.name}-tp"
+  region           = var.region
+  session_affinity = var.session_affinity
+
+  instances = var.instances
+
+  health_checks = [
+    google_compute_http_health_check.default.name,
+  ]
+}
+
+# ------------------------------------------------------------------------------
+# CREATE HEALTH CHECK
+# ------------------------------------------------------------------------------
+
+resource "google_compute_http_health_check" "default" {
+  project             = var.project
+  name                = "${var.name}-hc"
+  request_path        = var.health_check_path
+  port                = var.health_check_port
+  check_interval_sec  = var.health_check_interval
+  healthy_threshold   = var.health_check_healthy_threshold
+  unhealthy_threshold = var.health_check_unhealthy_threshold
+  timeout_sec         = var.health_check_timeout
+}
+
+# ------------------------------------------------------------------------------
+# CREATE FIREWALL FOR THE HEALTH CHECKS
+# ------------------------------------------------------------------------------
+
+# Health check firewall allows ingress tcp traffic from the health check IP addresses
+resource "google_compute_firewall" "health_check" {
+  project = var.network_project == "" ? var.project : var.network_project
```

Comment:
Should we use `null` for the default here?

Reply:
Is this something we would want to do consistently?

Reply:
I've been doing it consistently. It fits better the semantic when you pass through the vars to the resources, as it is the equivalent of omitting the var from the resource.

Reply:
Got it! Now using `null` as default.

Reply:
One gotcha with using `null` that I just ran into in `module-ecs` is that you can't interpolate it in a string. So so far, I've been using `null` if it allows, and `""` if the var is intended to be interpolated in a string (e.g `name_prefix`). But that seems a bit confusing... On the other hand, getting in the habit of using `null` is nice because it resolves bugs where we don't know what the default behavior of the provider is when an attribute is omitted...

What do you think here?